tests/platform_tests/fwutil/fwutil_common.py

```diff
@@ -134,20 +144,36 @@ def get_install_paths(duthost, fw, versions, chassis, target_component):
     for comp, revs in component.items():
         if comp in ver:
             if revs[0].get("upgrade_only", False) and ver[comp] not in [r["version"] for r in revs]:
-                log.warning("Firmware is upgrade only and existing firmware {} is not present in version list. Skipping {}".format(ver[comp], comp))
+                logger.warning("Firmware is upgrade only and existing firmware {} is not present in version list. "
+                               "Skipping {}".format(ver[comp], comp))
                 continue
             for i, rev in enumerate(revs):
                 if "hw_revision" in rev and rev["hw_revision"] != get_hw_revision(duthost):
-                    log.warning("Firmware {} only supports HW Revision {} and this chassis is {}. Skipping".format(rev["version"], rev["hw_revision"], get_hw_revision(duthost)))
+                    logger.warning("Firmware {} only supports HW Revision {} and this chassis is {}. Skipping".
+                                   format(rev["version"], rev["hw_revision"], get_hw_revision(duthost)))
                     continue
                 if rev["version"] != ver[comp]:
                     paths[comp] = rev
                     break
                 elif rev.get("upgrade_only", False):
-                    log.warning("Firmware is upgrade only and newer version than {} is not available. Skipping {}".format(ver[comp], comp))
+                    logger.warning("Firmware is upgrade only and newer version than {} is not available. Skipping {}".
+                                   format(ver[comp], comp))
                     break
     return paths
 
+
+def get_defined_components(duthost, defined_fw, chassis):
+    """
+    Update the component content if there is definition for the specific host.
+    Sometimes, if there is some host has specific component, they can have this defined in the firmware.json file
+    """
+    component = defined_fw["chassis"].get(chassis, {})["component"]
+    if duthost.hostname in defined_fw["host"]:
+        for component_type in defined_fw["host"][duthost.hostname]["component"].keys():
+            component[component_type] = defined_fw["host"][duthost.hostname]["component"][component_type]
+    return component
```

Comment:
why do we need to deep copy the sub-dictionary? can't we do it in a more pythonic way?